File: ml/analysis/verify_attention_forward.cpp
// Copyright 2025-2026 coRAN LABS Private Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "lib/scheduler/support/attention_scheduler_predictor.h"
#include <cstdio>

int main(int argc, char** argv)
{
  if (argc < 2) {
    std::printf("usage: verify <model.model>\n");
    return 1;
  }
  auto m = ocudu::attn_sched::load_model_file(argv[1]);
  if (!m) {
    std::printf("LOAD_FAILED\n");
    return 1;
  }

  const unsigned     k = 12;
  std::vector<float> phi(static_cast<size_t>(k) * m->n_feat);
  for (unsigned i = 0; i != k; ++i) {
    for (unsigned f = 0; f != m->n_feat; ++f) {
      phi[static_cast<size_t>(i) * m->n_feat + f] =
          0.1f * static_cast<float>(i + 1) - 0.3f * static_cast<float>(f);
    }
  }
  std::vector<float> c_vec(m->n_ctx);
  for (unsigned i = 0; i != m->n_ctx; ++i) {
    c_vec[i] = 0.5f + 0.25f * static_cast<float>(i);
  }
  std::vector<float>   delta(k);
  std::vector<uint8_t> mask(k, 1);
  for (unsigned i = 0; i != k; ++i) {
    delta[i] = 1.0f - 0.05f * static_cast<float>(i);
    if (i % 5 == 3) {
      mask[i] = 0;
    }
  }

  ocudu::attn_sched::workspace ws;
  ws.reserve(m->d_model, k);
  m->encode(phi.data(), k, ws);
  m->prepare_decode(ws);

  std::vector<float>       h_prev(m->d_model, 0.f);
  const std::vector<float> logp = m->decode_step(h_prev.data(), c_vec.data(), delta.data(), mask.data(), ws);
  for (unsigned i = 0; i != k; ++i) {
    std::printf("%.9f%s", logp[i], i + 1 == k ? "\n" : ",");
  }

  const int best = ocudu::attn_sched::model::argmax(logp);
  if (best >= 0) {
    const float* row = ws.h_enc.data() + static_cast<size_t>(best) * m->d_model;
    std::copy(row, row + m->d_model, h_prev.begin());
    mask[best] = 0;
  }
  const std::vector<float> logp2 = m->decode_step(h_prev.data(), c_vec.data(), delta.data(), mask.data(), ws);
  for (unsigned i = 0; i != k; ++i) {
    std::printf("%.9f%s", logp2[i], i + 1 == k ? "\n" : ",");
  }
  return 0;
}

File: ml/analysis/verify_attention_forward.py
# Copyright 2025-2026 coRAN LABS Private Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import sys

import numpy as np
import torch

sys.path.insert(0, "ml/training")
from attention_scheduler import AttentionScheduler

blob = json.load(open(sys.argv[1]))
model = AttentionScheduler(blob["n_feat"], blob["n_ctx"], blob["d_model"], blob["n_heads"]).double()

KEY = {
    "encoder.proj.weight": "enc_proj_w", "encoder.proj.bias": "enc_proj_b",
    "encoder.attn.in_proj_weight": "enc_attn_in_w", "encoder.attn.in_proj_bias": "enc_attn_in_b",
    "encoder.attn.out_proj.weight": "enc_attn_out_w", "encoder.attn.out_proj.bias": "enc_attn_out_b",
    "encoder.norm1.weight": "enc_norm1_w", "encoder.norm1.bias": "enc_norm1_b",
    "encoder.ff.0.weight": "enc_ff1_w", "encoder.ff.0.bias": "enc_ff1_b",
    "encoder.ff.2.weight": "enc_ff2_w", "encoder.ff.2.bias": "enc_ff2_b",
    "encoder.norm2.weight": "enc_norm2_w", "encoder.norm2.bias": "enc_norm2_b",
    "decoder.ctx.weight": "dec_ctx_w", "decoder.ctx.bias": "dec_ctx_b",
    "decoder.attn.in_proj_weight": "dec_attn_in_w", "decoder.attn.in_proj_bias": "dec_attn_in_b",
    "decoder.attn.out_proj.weight": "dec_attn_out_w", "decoder.attn.out_proj.bias": "dec_attn_out_b",
    "decoder.delta.weight": "dec_delta_w", "decoder.delta.bias": "dec_delta_b",
}
sd = {k: torch.tensor(np.array(blob[KEY[k]]).reshape(v.shape), dtype=torch.float64)
      for k, v in model.state_dict().items()}
model.load_state_dict(sd)
model.eval()

K = 12
n_feat, n_ctx, d = blob["n_feat"], blob["n_ctx"], blob["d_model"]
phi = np.array([[0.1 * (i + 1) - 0.3 * f for f in range(n_feat)] for i in range(K)])
c_vec = np.array([[0.5 + 0.25 * i for i in range(n_ctx)]])
delta = np.array([[1.0 - 0.05 * i for i in range(K)]])
mask = np.array([[(i % 5) != 3 for i in range(K)]])

with torch.no_grad():
    h_enc = model.encoder(torch.tensor(phi, dtype=torch.float64).unsqueeze(0))
    h_prev = torch.zeros(1, d, dtype=torch.float64)
    args = (h_enc, h_prev, torch.tensor(c_vec), torch.tensor(delta), torch.tensor(mask))
    logp = model.decoder(*args)[0].numpy()
    print(",".join(f"{v:.9f}" for v in logp))

    finite = np.where(np.isfinite(logp))[0]
    if len(finite):
        best = int(finite[np.argmax(logp[finite])])
        h_prev = h_enc[0, best].unsqueeze(0)
        mask[0, best] = False
    logp2 = model.decoder(h_enc, h_prev, torch.tensor(c_vec), torch.tensor(delta),
                          torch.tensor(mask))[0].numpy()
    print(",".join(f"{v:.9f}" for v in logp2))
